x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/AdjustableCapacityBlockingQueue.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.common;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+/**
+ * Provides a limited functionality queue that can have its capacity adjusted.
+ * @param <E> the items to store in the queue
+ */
+public class AdjustableCapacityBlockingQueue<E> {
+
+    private BlockingQueue<E> queue;
+    private final Function<Integer, BlockingQueue<E>> createQueue;
+    private final ReentrantReadWriteLock lock;
+
+    /**
+     * Constructs the adjustable capacity queue
+     * @param initialCapacity the initial capacity of the queue
+     * @param createQueue a function for creating a new backing queue with the specified capacity
+     */
+    public AdjustableCapacityBlockingQueue(int initialCapacity, Function<Integer, BlockingQueue<E>> createQueue) {
+        this.createQueue = Objects.requireNonNull(createQueue);
+        queue = createQueue.apply(initialCapacity);
+        lock = new ReentrantReadWriteLock();
+    }
+
+    /**
+     * Sets the capacity of the queue. If the new capacity is smaller than the current number of elements in the queue, the
+     * elements that exceed the new capacity size are returned and will not be present in the queue.
+     * This is potentially an expensive operation because a new internal queue is instantiated.
+     * @param capacity the new capacity for the queue.
+     * @return a list of elements that could not fit within the queue given the new capacity requirement.
+     */
+    public List<E> setCapacity(int capacity) {
+        final ReentrantReadWriteLock.WriteLock writeLock = lock.writeLock();
+        List<E> remainingItems = new ArrayList<>(queue.size());
+
+        writeLock.lock();
+        try {
+            BlockingQueue<E> newQueue = createQueue.apply(capacity);
+            queue.drainTo(newQueue, capacity);
+            queue.drainTo(remainingItems);
+            queue = newQueue;
+        } finally {
+            writeLock.unlock();
+        }
+
+        return remainingItems;
+    }
```

Comment:
Based on your explanation of why we return the excess items from the queue, I am left wondering why we need to create a new queue at all?

What if, instead of creating new queue objects and draining the queue elements into the new queue object, we instead handle all of the queue capacity logic in AdjustableCapacityBlockingQueue. That is, when the queue size is decreased, we don't reject the tasks in the queue, but rather we just stop accepting new ones as long as the size of the queue is greater than the new capacity.

As you mention, this is a bit of a corner case, so maybe its just not that important regardless, and your time would be better spent on something else. I only mention this because I imagine as a user I would be surprised if decreasing the queue size stopped running tasks. I would prefer to be able to adjust the queue size on the fly without worrying about restarting tasks.

Reply:
Great feedback Max, and good idea. I think I was able to accomplish this without too much complexity by introducing one more queue 😆 . Let me know what you think.